tokio-trace/README.md

```diff
@@ -0,0 +1,156 @@
+# tokio-trace
+
+A scoped, structured logging and diagnostics system.
+
+## Overview
+
+`tokio-trace` is a framework for instrumenting Rust programs to collect
+structured, event-based diagnostic information.
+
+In asynchronous systems like Tokio, interpreting traditional log messages can
+often be quite challenging. Since individual tasks are multiplexed on the same
+thread, associated events and log lines are intermixed making it difficult to
+trace the logic flow. `tokio-trace` expands upon logging-style diagnostics by
+allowing libraries and applications to record structured events with additional
+information about *temporality* and *causality* --- unlike a log message, a span
+in `tokio-trace` has a beginning and end time, may be entered and exited by the
+flow of execution, and may exist within a nested tree of similar spans. In
+addition, `tokio-trace` spans are *structured*, with the ability to record typed
+data as well as textual messages.
+
+The `tokio-trace` crate provides the APIs necessary for instrumenting libraries
+and applications to emit trace data.
+
+## Usage
+
+First, add this to your `Cargo.toml`:
+
+```toml
+[dependencies]
+tokio-trace = { git = "https://github.com/tokio-rs/tokio" }
+```
+
+Next, add this to your crate:
+
+```rust
+#[macro_use]
+extern crate tokio_trace;
+```
+
+This crate provides macros for creating `Span`s and `Event`s, which represent
```

Comment:
A useful thing to expand upon is when someone should use a `span` versus an `event`, and at what levels of granularity. A starting point would be something like:

- Use spans for things that correspond to a discrete unit of work (i.e., a given request's lifespan in a server with a quick request/response cycle) or use a span for each IO-action/each time you interact with a new external system (a database, a queue, whatever).
- Use events for things that happened _within_ a span—a request returned with a given status code, grabbed *n* items from a queue, etc.

Reply:
:+1: this is good advice, thanks --- do you think this kind of discussion is more appropriate for the README or for RustDoc?

Reply:
Both locations, I think, opting for greater detail in the Rustdoc.

Reply:
Added to the docs in 355a3d9 :)